src/graphics/shader-node-registry.js

```diff
@@ -0,0 +1,43 @@
+import { ShaderGraphNode } from '../scene/materials/shader-graph-node.js';
+
+/**
+ * @private
+ * @class
+ * @name pc.ShaderNodeRegistry
+ * @classdesc Container for all Shader Graph Core Nodes available to this application.
+ * @description Create an instance of a pc.ShaderNodeRegistry.
+ * @param {pc.Application} app - Application to attach registry to.
+ */
+function ShaderNodeRegistry(app) {
+    this.app = app;
+    this._nodes = { };
+    this._list = [];
+}
+
+ShaderNodeRegistry.prototype.constructor = ShaderNodeRegistry;
+
+ShaderNodeRegistry.prototype.destroy = function () {
+    this.app = null;
+};
+
+/**
+ * @private
+ * @function
+ * @name pc.ShaderNodeRegistry#register
+ * @description Register core nodes with registry.
+ * @param {object|string} coreNodeList - core node list object or JSON string
+ */
+ShaderNodeRegistry.prototype.register = function (coreNodeList) {
+    if (coreNodeList) {
+        if (typeof coreNodeList === 'string') {
+            coreNodeList = JSON.parse(coreNodeList);
+        }
+        var self = this;
+        Object.keys(coreNodeList).forEach(function (key) {
+            self._nodes[key] = new ShaderGraphNode(coreNodeList[key].code);
+            self._list.push(self._nodes[key]);
```

Comment:
We should filter the duplicates from the list. Perhaps with a warning in debug mode.

Reply:
I have not checked how this being used yet, just curious why the list is needed when you have dictionary like access with _nodes. Likely useful I'm sure, just checking.

Reply:
I will add a warning - the core nodes JSON asset should be considered an engine private asset and is invalid if core nodes are defined more than once - but being JSON format some developers will probably modify it - and a warning will also help engine development to identify and debug bad merges. The private _list is something I added for use (and is used) with the material editor prototype - and is quite useful for that.

Reply:
warning added